src/eredis_client.erl

```diff
@@ -529,64 +529,53 @@ transport_module(_) -> ssl.
+%% or {stop, ExitReason, State} like handle_info.
 maybe_reconnect(Reason, #state{reconnect_sleep = no_reconnect, queue = Queue} = State) ->
     reply_all({error, Reason}, Queue),
     %% If we aren't going to reconnect, then there is nothing else for
     %% this process to do.
-    {stop, normal, State#state{socket = undefined}};
+    {stop, normal, close_socket(State)};
+maybe_reconnect(Reason, #state{queue = Queue, reconnect_timer = TRef} = State)
+  when is_reference(TRef) ->
+    %% Reconnect already scheduled.
+    reply_all({error, Reason}, Queue),
+    {noreply, close_socket(State#state{queue = queue:new()})};
 maybe_reconnect(Reason,
                 #state{queue = Queue,
                        host = Host,
                        port = Port,
-                       socket_options = SocketOptions,
-                       tls_options = TlsOptions,
-                       connect_timeout = ConnectTimeout,
-                       reconnect_sleep = ReconnectSleep,
-                       auth_cmd = AuthCmd,
-                       database = Db} = State) ->
+                       reconnect_timer = undefined} = State) ->
     error_logger:error_msg("eredis: Re-establishing connection to ~p:~p due to ~p",
                            [Host, Port, Reason]),
-    Self = self(),
-    spawn_link(fun() ->
-                       process_flag(trap_exit, true),
-                       reconnect_loop(Self, ReconnectSleep, Host, Port,
-                                      SocketOptions, TlsOptions, ConnectTimeout,
-                                      AuthCmd, Db)
-               end),
-
-    %% tell all of our clients what has happened.
+    %% Tell all of our clients what has happened.
     reply_all({error, Reason}, Queue),
 
     %% Throw away the socket and the queue, as we will never get a
     %% response to the requests sent on the old socket. The absence of
     %% a socket is used to signal we are "down"
-    {noreply, State#state{socket = undefined, queue = queue:new()}}.
-
-%% @doc: Loop until a connection can be established, this includes
-%% successfully issuing the auth and select calls. When we have a
-%% connection, send the socket to Client in a message on the form
-%% `{connection_ready, Socket}'.
-reconnect_loop(Client, ReconnectSleep, Host, Port, SocketOptions,
-               TlsOptions, ConnectTimeout, AuthCmd, Db) ->
-    Client ! reconnect_attempt,
-    case connect(Host, Port, SocketOptions, TlsOptions, ConnectTimeout,
-                 AuthCmd, Db) of
-        {ok, Socket} ->
-            Client ! {connection_ready, Socket},
-            Transport = transport_module(TlsOptions),
-            Transport:controlling_process(Socket, Client),
-            Msgs = get_all_messages([]),
-            [Client ! M || M <- Msgs];
+    State1 = close_socket(State#state{queue = queue:new()}),
+    case connect(State1) of
+        {ok, State2} ->
+            {noreply, State2};
```

Comment:
Could we have
`schedule_reconnect(unknown, State2)`
here instead to match the error case?

Reply:
It is done inside connect/1 in the success case.

Reply:
I did it in connect in the success case as a side-effect that the caller doesn't even need to know.